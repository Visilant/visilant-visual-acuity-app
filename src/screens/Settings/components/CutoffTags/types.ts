import { VoidCallback } from '@components/types'
import { SnellenChartItem, SnellenChartRatio } from '@shared/domain/snellen-chart'
import { UnitSystem } from '@shared/domain/unit-system'
import { FunctionComponent } from 'react'

export type CutoffSelectorItem = SnellenChartItem & { selected: boolean }

export type CutoffSelectorPanel = {
  [key in SnellenChartRatio]: CutoffSelectorItem
}
export type UseCutoffTags = (
  value: SnellenChartRatio[],
  onChange: VoidCallback<SnellenChartRatio[]>
) => {
  toggleCutoff: VoidCallback<string>
  cutoffs: CutoffSelectorPanel
}

export interface CutoffTagsProps {
  onChange: VoidCallback<SnellenChartRatio[]>
  value: SnellenChartRatio[]
  units: UnitSystem
}

export type CutoffTagsComponent = FunctionComponent<CutoffTagsProps>
