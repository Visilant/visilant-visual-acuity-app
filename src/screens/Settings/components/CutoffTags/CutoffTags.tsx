import { CardTagRow, Tag, Text, TextStyles } from '@components'
import { useResultFormatter } from '@hooks/useResultFormatter'
import {
  sortByDecimal,
  useCutoffTags
} from '@screens/Settings/components/CutoffTags/cutoff-tags-utils'
import { CutoffTagsComponent } from '@screens/Settings/components/CutoffTags/types'
import React from 'react'

export const CutoffTags: CutoffTagsComponent = ({ value, onChange, units }) => {
  const { cutoffs, toggleCutoff } = useCutoffTags(value, onChange)
  const { format, base } = useResultFormatter(units)

  const cutoffTags = Object.entries(cutoffs)
    .sort(sortByDecimal)
    .map(([key, value]) => (
      <Tag
        key={key}
        text={`${base}/${format(value.ratio)}`}
        selected={value.selected}
        onPress={() => toggleCutoff(key)}
      />
    ))

  return (
    <>
      <CardTagRow>
        <Text namespace="screens" text="settings.cutoffs" textStyle={TextStyles.Label} />
      </CardTagRow>
      <CardTagRow>{cutoffTags}</CardTagRow>
    </>
  )
}
