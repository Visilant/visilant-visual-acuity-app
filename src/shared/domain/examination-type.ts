import { EyesResult } from './eyes-result'

export enum ExaminationType {
  NormalRight = 'NormalRight',
  NormalLeft = 'NormalLeft',
  PinholeRight = 'PinholeRight',
  PinholeLeft = 'PinholeLeft'
}

export type ExaminationData = {
  patientName?: string
  comment?: string
  normal?: EyesResult
  pinhole?: EyesResult
}

export interface ExaminationResult extends ExaminationData {
  mobileCreatedAt?: Date
}

export interface ExaminationToSave extends ExaminationData {
  mobileId?: string
}
