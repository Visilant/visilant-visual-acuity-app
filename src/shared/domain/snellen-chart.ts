/*
20ft    10ft    6m    3m    dec   mar  logmar
20/1000	10/500	6/300	3/150	0.02	50	  1.70
20/800	10/400	6/240	3/120	0.025	40	  1.60
20/600	10/300	6/180	3/90	0.033	30	  1.48
20/500	10/250	6/150	3/75	0.04	25	  1.40
20/400	10/200	6/120	3/60	0.05	20	  1.30
20/300	10/150	6/90	3/45	0.067	15	  1.18
20/250	10/125	6/75	3/37	0.08	12.5	1.10
20/200	10/100	6/60	3/30	0.10	10	  1.00
20/160	10/80	  6/48	3/24	0.125	8	    0.90
20/125	10/62	  6/38	3/19	0.16	6.25	0.80
20/100	10/50	  6/30	3/15	0.20	5	    0.70
20/80	  10/40	  6/24	3/12	0.25	4	    0.60
20/60	  10/30	  6/18	3/9	  0.33	3	    0.48
20/50	  10/25	  6/15	3/7.5	0.40	2.5	  0.40
20/40	  10/20	  6/12	3/6	  0.50	2	    0.30
20/30	  10/15	  6/9	  3/4.5	0.67	1.5	  0.18
20/25	  10/12	  6/7.5	3/4	  0.80	1.25	0.10
20/20	  10/10	  6/6	  3/3	  1.00	1	    0.00
20/16	  10/8	  6/4.8	3/2.4	1.25	0.8	  −0.10
20/12.5	10/6	  6/3.8	3/2	  1.60	0.625	−0.20
20/10	  10/5	  6/3	  3/1.5	2.00	0.5	  −0.30
20/8	  10/4	  6/2.4	3/1.2	2.50	0.4	  −0.40
20/6.6	10/3.3	6/2	  3/1	  3.00	0.333	−0.48

distance/acuity=dec acuity/distance=mar log(mar)=logmar

*/

import { UnitSystem } from './unit-system'

export interface SnellenChartItem {
  ratio: SnellenChartRatio
  metric: number
  imperial: number
  decimal: number
}

export const getMar = (value: SnellenChartItem) => 1 / value.decimal

export const getLogMar = (value: SnellenChartItem) => Math.log10(1 / value.decimal)

export type SnellenDictionary = {
  [key in SnellenChartRatio]: SnellenChartItem
}

export const SnellenChartDictionary: SnellenDictionary = {
  '0.1': {
    ratio: '0.1',
    metric: 60,
    imperial: 200,
    decimal: 0.1
  },
  '0.125': {
    ratio: '0.125',
    metric: 48,
    imperial: 160,
    decimal: 0.125
  },
  '0.16': {
    ratio: '0.16',
    metric: 38,
    imperial: 125,
    decimal: 0.16
  },
  '0.2': {
    ratio: '0.2',
    metric: 30,
    imperial: 100,
    decimal: 0.2
  },
  '0.25': {
    ratio: '0.25',
    metric: 24,
    imperial: 80,
    decimal: 0.25
  },
  '0.33': {
    ratio: '0.33',
    metric: 18,
    imperial: 60,
    decimal: 0.33
  },
  '0.4': {
    ratio: '0.4',
    metric: 15,
    imperial: 50,
    decimal: 0.4
  },
  '0.5': {
    ratio: '0.5',
    metric: 12,
    imperial: 40,
    decimal: 0.5
  },
  '0.67': {
    ratio: '0.67',
    metric: 9,
    imperial: 30,
    decimal: 0.67
  },
  '0.8': {
    ratio: '0.8',
    metric: 7.5,
    imperial: 25,
    decimal: 0.8
  },
  '1.0': {
    ratio: '1.0',
    metric: 6,
    imperial: 20,
    decimal: 1
  }
}

export type SnellenChartRatio =
  | '0.1'
  | '0.125'
  | '0.16'
  | '0.2'
  | '0.25'
  | '0.33'
  | '0.4'
  | '0.5'
  | '0.67'
  | '0.8'
  | '1.0'
export type BaseDistanceType = {
  [key in UnitSystem]: number
}

export const snellenChartList = Object.keys(SnellenChartDictionary) as SnellenChartRatio[]

export const BaseDistance: BaseDistanceType = {
  Metric: 6,
  Imperial: 20
}

export const isSnellenChartRatio = (test: string): test is SnellenChartRatio => {
  return SnellenChartDictionary[test] !== undefined
}

export const sortByDecimal = (a, b) =>
  SnellenChartDictionary[a].decimal - SnellenChartDictionary[b].decimal

export const sortSnellenChartRatios = (ratios: SnellenChartRatio[]): SnellenChartRatio[] => {
  return ratios.sort(sortByDecimal)
}
