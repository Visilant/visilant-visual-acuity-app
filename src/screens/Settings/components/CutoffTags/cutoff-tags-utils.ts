import { SnellenChartDictionary, SnellenChartRatio } from '@shared/domain/snellen-chart'
import {
  CutoffSelectorItem,
  CutoffSelectorPanel,
  UseCutoffTags
} from '@screens/Settings/components/CutoffTags/types'
import { useMemo } from 'react'

export const sortByDecimal = (a: [string, CutoffSelectorItem], b: [string, CutoffSelectorItem]) =>
  b[1].decimal - a[1].decimal

export const createCutoffs = (selectedCutoffs: SnellenChartRatio[]): CutoffSelectorPanel => {
  const result = { ...SnellenChartDictionary }

  for (const key in result) {
    const item = result[key]

    const snellenKey = key as SnellenChartRatio
    const selected = selectedCutoffs.includes(snellenKey)

    item.selected = selected
  }

  return result as CutoffSelectorPanel
}

export const useCutoffTags: UseCutoffTags = (value, onChange) => {
  const toggleCutoff = (key: string) => {
    const target = key as SnellenChartRatio

    let newValue: SnellenChartRatio[]

    if (value.includes(target)) {
      newValue = value.filter(item => item !== target)
    } else {
      newValue = [...value, target]
    }

    onChange(newValue)
  }

  const cutoffs = useMemo(() => createCutoffs(value), [value])

  return { cutoffs, toggleCutoff }
}
