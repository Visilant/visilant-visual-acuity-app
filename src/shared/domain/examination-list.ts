import { Examination } from './examination'

export class ExaminationList {
  constructor(private _values: Examination[]) {}
  get values(): Examination[] {
    return this._values
  }
}
