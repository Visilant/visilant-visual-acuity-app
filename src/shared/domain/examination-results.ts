import { EyeResult, EyesResult } from '@shared/domain/eyes-result'
import { ExaminationType } from '@shared/domain/examination-type'

export class ExaminationResults {
  constructor(
    private _mobileId?: string,
    private _normal?: EyesResult,
    private _pinhole?: EyesResult
  ) {}

  get normal(): EyesResult | undefined {
    return this._normal
  }

  get pinhole(): EyesResult | undefined {
    return this._pinhole
  }
  get mobileId(): string | undefined {
    return this._mobileId
  }

  copy(): ExaminationResults {
    return new ExaminationResults(this._mobileId, this._normal, this._pinhole)
  }

  public setSnellen(type: ExaminationType, value: EyeResult | undefined) {
    if (type === ExaminationType.NormalLeft) {
      this._normal = EyesResult.withLeft(this._normal, value)
    } else if (type === ExaminationType.NormalRight) {
      this._normal = EyesResult.withRight(this._normal, value)
    } else if (type === ExaminationType.PinholeLeft) {
      this._pinhole = EyesResult.withLeft(this._pinhole, value)
    } else if (type === ExaminationType.PinholeRight) {
      this._pinhole = EyesResult.withRight(this._pinhole, value)
    }
  }
}
