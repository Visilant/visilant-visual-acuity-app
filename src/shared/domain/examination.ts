import { EyesResult } from '@shared/domain/eyes-result'
import { ExaminationResult } from '@shared/domain/examination-type'
import { UUID } from '@shared/uuid'

export class Examination {
  constructor(
    private _mobileId: string,
    private _patientName: string,
    private _comment: string,
    private _mobileCreatedAt: Date,
    private _normal: EyesResult | undefined,
    private _pinhole: EyesResult | undefined,
    private _isSynced: boolean = true
  ) {}

  static fromResults(results: ExaminationResult): Examination {
    const { patientName, comment, mobileCreatedAt, normal, pinhole } = results
    return new Examination(
      UUID.v4(),
      patientName || '',
      comment || '',
      mobileCreatedAt || new Date(),
      normal,
      pinhole,
      false
    )
  }

  get mobileId(): string {
    return this._mobileId
  }

  get patientName(): string {
    return this._patientName
  }

  get comment(): string {
    return this._comment
  }

  get normal(): EyesResult | undefined {
    return this._normal
  }

  get pinhole(): EyesResult | undefined {
    return this._pinhole
  }

  get mobileCreatedAt(): Date {
    return this._mobileCreatedAt
  }

  get isSynced(): boolean {
    return this._isSynced
  }
}
