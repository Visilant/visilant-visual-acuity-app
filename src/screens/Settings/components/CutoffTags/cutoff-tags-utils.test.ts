import { createCutoffs } from '@screens/Settings/components/CutoffTags/cutoff-tags-utils'
import { SnellenChartRatio } from '@shared/domain/snellen-chart'

test('createCutoffs', () => {
  const testValue: SnellenChartRatio[] = ['0.1', '1.0', '0.25']

  const result = createCutoffs(testValue)
  expect(result).toMatchSnapshot()
})
