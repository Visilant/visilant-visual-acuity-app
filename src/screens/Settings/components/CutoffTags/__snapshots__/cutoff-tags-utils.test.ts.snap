// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCutoffs 1`] = `
{
  "0.1": {
    "decimal": 0.1,
    "imperial": 200,
    "metric": 60,
    "ratio": "0.1",
    "selected": true,
  },
  "0.125": {
    "decimal": 0.125,
    "imperial": 160,
    "metric": 48,
    "ratio": "0.125",
    "selected": false,
  },
  "0.16": {
    "decimal": 0.16,
    "imperial": 125,
    "metric": 38,
    "ratio": "0.16",
    "selected": false,
  },
  "0.2": {
    "decimal": 0.2,
    "imperial": 100,
    "metric": 30,
    "ratio": "0.2",
    "selected": false,
  },
  "0.25": {
    "decimal": 0.25,
    "imperial": 80,
    "metric": 24,
    "ratio": "0.25",
    "selected": true,
  },
  "0.33": {
    "decimal": 0.33,
    "imperial": 60,
    "metric": 18,
    "ratio": "0.33",
    "selected": false,
  },
  "0.4": {
    "decimal": 0.4,
    "imperial": 50,
    "metric": 15,
    "ratio": "0.4",
    "selected": false,
  },
  "0.5": {
    "decimal": 0.5,
    "imperial": 40,
    "metric": 12,
    "ratio": "0.5",
    "selected": false,
  },
  "0.67": {
    "decimal": 0.67,
    "imperial": 30,
    "metric": 9,
    "ratio": "0.67",
    "selected": false,
  },
  "0.8": {
    "decimal": 0.8,
    "imperial": 25,
    "metric": 7.5,
    "ratio": "0.8",
    "selected": false,
  },
  "1.0": {
    "decimal": 1,
    "imperial": 20,
    "metric": 6,
    "ratio": "1.0",
    "selected": true,
  },
}
`;
